libpod/boltdb_state.go

```diff
@@ -73,6 +75,12 @@ func NewBoltState(path, lockDir string, runtime *Runtime) (State, error) {
 	if err != nil {
 		return nil, errors.Wrapf(err, "error opening database %s", path)
 	}
+	// Everywhere else, we use s.closeDBCon(db) to ensure the state's DB
+	// mutex is also unlocked.
+	// However, here, the mutex has not been locked, since we just created
+	// the DB connection, and it hasn't left this function yet - no risk of
+	// concurrent access.
```

Comment:
This general approach seems broken to me.  Won't this race with other parallel Bolt connections from this process?  With unreliable flocking and no shared mutex, I don't see a way to block that race.

It seems like what we need is a singleton (per path?).  Then anyone who needs it grabs the singleton and tries to acquire the mutex (to protect against intra-process races), and then tries to aquire the flock (to protect against inter-process races).  Once you're done, you could release the flock and then release the mutex.  Or to be fancy in cases where you had other goroutines already queued on the mutex, you might be able to hold the flock and only release the mutex.  But I think we need to ensure that each DB path has one, and only one, mutex at a time for intra-process synchronization.

Reply:
This is the initial setup code for the BoltDB state - it should only run once per libpod instance, before the state is made available for use. So we don't need to worry about races, as the DB will not have users until after this function returns.

Reply:
> This is the initial setup code for the BoltDB state - it should only run once per libpod instance, before the state is made available for use.

Ah, so it's effectively a singleton already.  Maybe update [the function docs][1] to point out that it should only ever be called once for a given `path`, because parallel instances on the same path within the same process will not be protected from racy access?  And a simple guard for that would be to have a package global:

```go
var boltPaths map[string]interface  // or whatever value type is idiomatic Go for "we don't care about the values; this is just a set of keys"
var newBoltMutex sync.Mutex
```

and then a guard in `NewBoltState` like:

```go
newBoltMutex.Lock()
defer newBoltMutex.Unlock()
_, ok := boltPaths[path]
if ok {
  return nil, fmt.Errorf("NewBoltState has already been called for %q", path)
}
boltPaths[path] = nil  // with whatever dummy value is appropriate for the boltPaths value type
```

Alternatively, instead of making it an error to ask twice, `boltPaths` could store the lock dir, runtime, and resulting `State`, and just return the pre-existing `State` if a second request with the same `path`, `lockDir`, and `runtime` was received.  Since matching `runtime` seems tedious, it's probably easier to just error with something like my guard above.

[1]: https://github.com/projectatomic/libpod/blob/8ce0e0b2460220fccbb31fced940604b55d8a195/libpod/boltdb_state.go#L52